core/src/stored/backends/dplcompat_device.cc

```diff
@@ -0,0 +1,418 @@
+          T_("Failed to read %s (%ld) to big to fit in chunksize of %ld "
+             "bytes\n"),
+          obj_name.c_str(), obj_stat->size, request->wbuflen);
+    Dmsg1(debug_info, "%s", errmsg);
+    dev_errno = EINVAL;
+    return false;
+  }
+
+  if (auto obj_data = m_storage.download(obj_name, obj_chunk,
+                                         {request->buffer, obj_stat->size})) {
+    *request->rbuflen = obj_data->size_bytes();
+    return true;
+  } else {
+    PmStrcpy(errmsg, obj_data.error().c_str());
+    Dmsg1(debug_info, "%s", errmsg);
+    dev_errno = EIO;
+    return false;
+  }
+}
+
+bool DropletCompatibleDevice::TruncateRemoteVolume(DeviceControlRecord*)
+{
+  const char* vol_name = getVolCatName();
+  const auto list = m_storage.list(vol_name);
+  if (!list) {
+    PmStrcpy(errmsg, list.error().c_str());
+    dev_errno = EIO;
+    return false;
+  }
+  for (const auto& [chunk_name, stat] : *list) {
+    if (is_chunk_name(chunk_name)) {
+      if (auto res = m_storage.remove(vol_name, chunk_name); !res) {
+        PmStrcpy(errmsg, list.error().c_str());
+        dev_errno = EIO;
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+ssize_t DropletCompatibleDevice::RemoteVolumeSize()
+{
+  const auto list = m_storage.list(getVolCatName());
+  if (!list) {
+    PmStrcpy(errmsg, list.error().c_str());
+    dev_errno = EIO;
+    return false;
+  }
+  if (list->empty()) { return -1; }
+  ssize_t total_size{0};
+  for (const auto& [name, stat] : *list) {
+    if (is_chunk_name(name)) { total_size += stat.size; }
+  }
+  return total_size;
+}
+
+
+bool DropletCompatibleDevice::d_flush(DeviceControlRecord*)
+{
+  return WaitUntilChunksWritten();
+};
+
+int DropletCompatibleDevice::d_open(const char* pathname, int flags, int mode)
+{
+  if (!setup()) {
+    dev_errno = EIO;
+    Emsg0(M_FATAL, 0, errmsg);
+  }
+  return SetupChunk(pathname, flags, mode);
+}
+
+ssize_t DropletCompatibleDevice::d_read(int t_fd, void* buffer, size_t count)
+{
+  return ReadChunked(t_fd, buffer, count);
+}
+
+ssize_t DropletCompatibleDevice::d_write(int t_fd,
+                                         const void* buffer,
+                                         size_t count)
+{
+  return WriteChunked(t_fd, buffer, count);
+}
+
+int DropletCompatibleDevice::d_close(int) { return CloseChunk(); }
+
+int DropletCompatibleDevice::d_ioctl(int, ioctl_req_t, char*) { return -1; }
+
+boffset_t DropletCompatibleDevice::d_lseek(DeviceControlRecord*,
+                                           boffset_t offset,
+                                           int whence)
+{
+  switch (whence) {
+    case SEEK_SET:
+      offset_ = offset;
+      break;
+    case SEEK_CUR:
+      offset_ += offset;
+      break;
+    case SEEK_END: {
+      ssize_t volumesize;
+
+      volumesize = ChunkedVolumeSize();
+
+      Dmsg1(debug_info, "Current volumesize: %lld\n", volumesize);
+
+      if (volumesize >= 0) {
+        offset_ = volumesize + offset;
```

Comment:
We should check that `offset` is `<= 0` but `>= -volumesize`, otherwise this goes "out of bounds".

Reply:
But then we'd need to add that check for SEEK_SET and SEEK_CUR, too. Which would require a call to `ChunkedVolumeSize()` for every `d_lseek()` instead of only when `whence == SEEK_END`.
I'm not sure that this is worth doing.

Reply:
Makes sense